util/ribbon_impl.h

```diff
@@ -0,0 +1,487 @@
+  // false. Caller guarantees that the number of inputs in this batch
+  // does not exceed `backtrack_size` provided to Reset.
+  //
+  // Adding can fail even before all the "slots" are completely "full".
+  //
+  template <typename InputIterator>
+  bool AddRangeOrRollBack(InputIterator begin, InputIterator end) {
+    return BandingAddRange(this, this, *this, begin, end);
+  }
+
+  // Adds a single input to the banding, returning true if successful.
+  // If unsuccessful, returns false and banding state is unchanged.
+  //
+  // Adding can fail even before all the "slots" are completely "full".
+  //
+  bool Add(const AddInput& input) { return AddRange(&input, &input + 1); }
+
+  // Return the number of "occupied" rows (with non-zero coefficients stored).
+  Index GetOccupiedCount() const {
+    Index count = 0;
+    const Index num_slots = num_starts_ + kCoeffBits - 1;
+    for (Index i = 0; i < num_slots; ++i) {
+      if (coeff_rows_[i] != 0) {
+        ++count;
+      }
+    }
+    return count;
+  }
+
+  // ********************************************************************
+  // High-level API
+
+  // Iteratively (a) resets the structure for `num_slots`, (b) attempts
+  // to add the range of inputs, and (c) if unsuccessful, chooses next
+  // hash seed, until either successful or unsuccessful with max_seed
+  // (minimum one seed attempted). Returns true if successful. In that
+  // case, use GetSeed() to get the successful seed.
+  template <typename InputIterator>
+  bool ResetAndFindSeedToSolve(Index num_slots, InputIterator begin,
+                               InputIterator end, Seed max_seed) {
+    StandardHasher<TypesAndSettings>::ResetSeed();
+    do {
+      Reset(num_slots);
+      bool success = AddRange(begin, end);
+      if (success) {
+        return true;
+      }
+    } while (StandardHasher<TypesAndSettings>::NextSeed(max_seed));
+    // no seed through max_seed worked
+    return false;
+  }
+
+ protected:
+  // TODO: explore combining in a struct
+  std::unique_ptr<CoeffRow[]> coeff_rows_;
+  std::unique_ptr<ResultRow[]> result_rows_;
+  Index num_starts_ = 0;
+  Index num_slots_allocated_ = 0;
+  std::unique_ptr<Index[]> backtrack_;
+  Index backtrack_size_ = 0;
+};
+
+// Implements concept SimpleSolutionStorage, mostly for demonstration
+// purposes. This is "in memory" only because it does not handle byte
+// ordering issues for serialization.
+template <class TypesAndSettings>
+class InMemSimpleSolution {
+ public:
+  IMPORT_RIBBON_TYPES_AND_SETTINGS(TypesAndSettings);
+
+  void PrepareForNumStarts(Index num_starts) {
+    const Index num_slots = num_starts + kCoeffBits - 1;
+    assert(num_slots >= kCoeffBits);
+    if (num_slots > num_slots_allocated_) {
+      // Do not need to init the memory
+      solution_rows_.reset(new ResultRow[num_slots]);
+      num_slots_allocated_ = num_slots;
+    }
+    num_starts_ = num_starts;
+  }
+
+  Index GetNumStarts() const { return num_starts_; }
```

Comment:
I should add a comment for that. It's because num_starts is on the extreme critical path for queries: it's value is needed (for FastRange) to know where to start probing in memory. We don't want any extra instructions for getting that value.